File: docs/calculator.js
// Calculator functionality
class PaymentCalculator {
    constructor() {
        this.form = document.getElementById('calculatorForm');
        this.resultsContent = document.getElementById('resultsContent');
        this.detailsCard = document.getElementById('detailsCard');
        this.paymentSchedule = document.getElementById('paymentSchedule');
        this.calculationExplanation = document.getElementById('calculationExplanation');
        
        this.init();
    }
    
    init() {
        // Form submission
        this.form.addEventListener('submit', (e) => {
            e.preventDefault();
            this.calculate();
        });
        
        // Real-time calculations
        const inputs = this.form.querySelectorAll('input');
        inputs.forEach(input => {
            input.addEventListener('input', () => {
                this.calculate();
            });
        });
        
        // Initial calculation
        this.calculate();
    }
    
    getFormData() {
        const formData = new FormData(this.form);
        return {
            totalAmount: parseFloat(formData.get('totalAmount')) || 0,
            installments: parseInt(formData.get('installments')) || 0,
            onTimeFee: parseFloat(formData.get('onTimeFee')) || 0,
            returnRate: parseFloat(formData.get('returnRate')) || 0,
            firstPaymentDate: formData.get('firstPaymentDate'),
            lumpSumDate: formData.get('lumpSumDate')
        };
    }
    
    addMonths(date, months) {
        const result = new Date(date);
        result.setMonth(result.getMonth() + months);
        return result;
    }
    
    differenceInMonths(date1, date2) {
        const d1 = new Date(date1);
        const d2 = new Date(date2);
        return (d1.getFullYear() - d2.getFullYear()) * 12 + (d1.getMonth() - d2.getMonth());
    }
    
    formatDate(date) {
        return date.toLocaleDateString('en-US', { 
            month: 'short', 
            day: 'numeric', 
            year: 'numeric' 
        });
    }
    
    calculatePaymentStrategy(data) {
        const { totalAmount, installments, onTimeFee, returnRate, firstPaymentDate, lumpSumDate } = data;
        
        // Validation
        if (!totalAmount || !installments || !firstPaymentDate || !lumpSumDate) {
            return null;
        }
        
        const firstDate = new Date(firstPaymentDate);
        const lumpDate = new Date(lumpSumDate);
        const installmentAmount = totalAmount / installments;
        
        // Generate payment schedule
        const paymentSchedule = [];
        for (let i = 0; i < installments; i++) {
            const paymentDate = this.addMonths(firstDate, i);
            const remainingBalance = totalAmount - (installmentAmount * (i + 1));
            paymentSchedule.push({
                paymentNumber: i + 1,
                date: this.formatDate(paymentDate),
                amount: installmentAmount,
                fee: i === 0 ? onTimeFee : 0, // One-time fee only on first payment
                remainingBalance: Math.max(0, remainingBalance)
            });
        }
        
        // Calculate installment strategy
        const totalFees = onTimeFee; // One-time fee for the entire installment plan
        
        // Calculate investment returns for remaining balance with monthly compounding
        let investmentReturns = 0;
        const monthlyReturnRate = returnRate / 100 / 12; // Monthly return rate
        const investmentBreakdown = [];
        
        for (let i = 0; i < installments - 1; i++) {
            const remainingBalance = totalAmount - (installmentAmount * (i + 1));
            if (remainingBalance > 0) {
                const monthlyReturn = remainingBalance * monthlyReturnRate;
                investmentReturns += monthlyReturn;
                investmentBreakdown.push({
                    month: i + 1,
                    balance: remainingBalance,
                    monthlyReturn: monthlyReturn,
                    cumulativeReturn: investmentReturns
                });
            }
        }
        
        const installmentNetCost = totalAmount + totalFees - investmentReturns;
        
        // Calculate lump sum strategy
        const monthsToLumpSum = this.differenceInMonths(lumpDate, firstDate);
        const opportunityCost = totalAmount * (returnRate / 100) * (monthsToLumpSum / 12);
        const lumpSumNetCost = totalAmount + opportunityCost;
        
        // Determine recommendation
        const savings = Math.abs(installmentNetCost - lumpSumNetCost);
        const recommendation = {
            strategy: lumpSumNetCost < installmentNetCost ? 'Lump Sum' : 'Installment',
            savings: savings
        };
        
        return {
            recommendation,
            installment: {
                totalAmount,
                totalFees,
                investmentReturns,
                netCost: installmentNetCost
            },
            lumpSum: {
                totalAmount,
                fees: 0,
                opportunityCost,
                netCost: lumpSumNetCost
            },
            paymentSchedule,
            investmentBreakdown
        };
    }
    
    calculate() {
        const data = this.getFormData();
        const results = this.calculatePaymentStrategy(data);
        
        if (results) {
            this.displayResults(results, data);
            this.detailsCard.style.display = 'block';
        } else {
            this.displayEmptyState();
            this.detailsCard.style.display = 'none';
        }
    }
    
    displayEmptyState() {
        this.resultsContent.innerHTML = `
            <div class="empty-state">
                <div class="empty-icon">🧮</div>
                <p>Enter your payment details to see the comparison</p>
            </div>
        `;
    }
    
    displayResults(results, data) {
        // Main results
        this.resultsContent.innerHTML = `
            <div class="recommendation">
                <div class="recommendation-content">
                    <div class="recommendation-left">
                        <div>👍</div>
                        <div>
                            <div class="recommendation-title">Recommended Strategy</div>
                            <div class="recommendation-subtitle">Based on your inputs</div>
                        </div>
                    </div>
                    <div class="recommendation-right">
                        <div class="recommendation-strategy">${results.recommendation.strategy}</div>
                        <div class="recommendation-savings">
                            Saves <span class="savings-amount">$${results.recommendation.savings.toFixed(0)}</span>
                        </div>
                    </div>
                </div>
            </div>
            
            <div class="comparison-cards">
                <div class="comparison-card">
                    <div class="comparison-header">
                        <div class="comparison-title">💳 Installment Payment Plan</div>
                    </div>
                    <div class="comparison-content">
                        <div class="comparison-row">
                            <span class="comparison-label">Total Amount</span>
                            <span class="comparison-value">$${results.installment.totalAmount.toLocaleString()}</span>
                        </div>
                        <div class="comparison-row">
                            <span class="comparison-label">One-Time Fee</span>
                            <span class="comparison-value text-accent">$${results.installment.totalFees.toLocaleString()}</span>
                        </div>
                        <div class="comparison-row">
                            <span class="comparison-label">Investment Returns</span>
                            <span class="comparison-value text-secondary">$${results.installment.investmentReturns.toFixed(0)}</span>
                        </div>
                        <div class="comparison-row net-cost-row">
                            <span class="comparison-label font-semibold">Net Cost</span>
                            <span class="comparison-value net-cost-value">$${results.installment.netCost.toFixed(0)}</span>
                        </div>
                    </div>
                </div>
                
                <div class="comparison-card">
                    <div class="comparison-header">
                        <div class="comparison-title">💵 Lump Sum Payment</div>
                    </div>
                    <div class="comparison-content">
                        <div class="comparison-row">
                            <span class="comparison-label">Total Amount</span>
                            <span class="comparison-value">$${results.lumpSum.totalAmount.toLocaleString()}</span>
                        </div>
                        <div class="comparison-row">
                            <span class="comparison-label">One-Time Fee</span>
                            <span class="comparison-value text-secondary">$${results.lumpSum.fees}</span>
                        </div>
                        <div class="comparison-row">
                            <span class="comparison-label">Opportunity Cost</span>
                            <span class="comparison-value text-accent">$${results.lumpSum.opportunityCost.toFixed(0)}</span>
                        </div>
                        <div class="comparison-row net-cost-row">
                            <span class="comparison-label font-semibold">Net Cost</span>
                            <span class="comparison-value net-cost-value">$${results.lumpSum.netCost.toFixed(0)}</span>
                        </div>
                    </div>
                </div>
            </div>
        `;
        
        // Payment schedule
        this.displayPaymentSchedule(results.paymentSchedule);
        
        // Calculation explanation
        this.displayCalculationExplanation(results, data);
    }
    
    displayPaymentSchedule(schedule) {
        this.paymentSchedule.innerHTML = schedule.map(payment => `
            <div class="payment-item">
                <div class="payment-info">
                    <span class="payment-number">Payment ${payment.paymentNumber}</span>
                    <span class="payment-date">${payment.date}</span>
                </div>
                <div class="payment-amount">
                    <div class="payment-value">$${payment.amount.toLocaleString()}</div>
                    ${payment.fee > 0 ? `<div class="payment-fee">+$${payment.fee} fee</div>` : ''}
                </div>
            </div>
        `).join('');
    }
    
    displayCalculationExplanation(results, data) {
        const investmentBreakdownHtml = results.investmentBreakdown.length > 0 ? `
            <div class="investment-breakdown">
                <div class="breakdown-title">Monthly Investment Returns (Compound Interest)</div>
                <div class="breakdown-items">
                    ${results.investmentBreakdown.map(item => `
                        <div class="breakdown-item">
                            <span class="font-medium">Month ${item.month}</span>
                            <div class="breakdown-calculation">
                                <div>$${item.balance.toLocaleString()} × ${(data.returnRate / 12).toFixed(2)}%</div>
                                <div class="breakdown-result">= $${item.monthlyReturn.toFixed(2)}</div>
                            </div>
                        </div>
                    `).join('')}
                    <div class="breakdown-item breakdown-total">
                        <span>Total Investment Returns</span>
                        <span>$${results.installment.investmentReturns.toFixed(2)}</span>
                    </div>
                </div>
            </div>
        ` : '';
        
        this.calculationExplanation.innerHTML = `
            <div class="explanation-section installment-section">
                <div class="explanation-title text-primary">Installment Strategy</div>
                <ul class="explanation-list">
                    <li>• Total payments: $${results.installment.totalAmount.toLocaleString()}</li>
                    <li>• One-time fee: $${results.installment.totalFees.toLocaleString()}</li>
                    <li>• Investment returns: $${results.installment.investmentReturns.toFixed(0)}</li>
                    <li>• <strong>Net cost: $${results.installment.netCost.toFixed(0)}</strong></li>
                </ul>
                ${investmentBreakdownHtml}
            </div>
            
            <div class="explanation-section lumpsum-section">
                <div class="explanation-title text-secondary">Lump Sum Strategy</div>
                <ul class="explanation-list">
                    <li>• Total payment: $${results.lumpSum.totalAmount.toLocaleString()}</li>
                    <li>• One-time fee: $${results.lumpSum.fees}</li>
                    <li>• Opportunity cost: $${results.lumpSum.opportunityCost.toFixed(0)}</li>
                    <li>• <strong>Net cost: $${results.lumpSum.netCost.toFixed(0)}</strong></li>
                </ul>
            </div>
            
            <div class="explanation-section insight-section">
                <div class="insight-header">
                    <span>💡</span>
                    <span class="font-medium">Key Insight</span>
                </div>
                <p class="insight-text">
                    The ${results.recommendation.strategy.toLowerCase()} payment saves $${results.recommendation.savings.toFixed(0)} 
                    ${results.recommendation.strategy === 'Lump Sum' 
                        ? 'because the avoided fees exceed the opportunity cost of not investing the money.'
                        : 'because the investment returns on the remaining balance exceed the additional fees paid.'
                    }
                </p>
            </div>
        `;
    }
}

// Initialize calculator when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    new PaymentCalculator();
});

File: tuition_strategy/calculator.py
import calendar

from datetime import datetime
from dataclasses import dataclass
from typing import Any, Dict, List

@dataclass
class PaymentScheduleItem:
    payment_number: int
    date: str
    amount: float
    fee: float
    remaining_balance: float
    accumulated_payments: int = 0


@dataclass
class InvestmentBreakdownItem:
    month: int
    balance: float
    monthly_return: float
    cumulative_return: float


@dataclass
class Results:
    recommendation: Dict[str, Any]
    installment: Dict[str, float]
    lump_sum: Dict[str, float]
    payment_schedule: List[PaymentScheduleItem]
    investment_breakdown: List[InvestmentBreakdownItem]


class PaymentCalculator:
    @staticmethod
    def add_months(date: datetime, months: int) -> datetime:
        """Add months to a date"""
        month = date.month - 1 + months
        year = date.year + month // 12
        month = month % 12 + 1
        day = min(date.day, calendar.monthrange(year, month)[1])
        return datetime(year, month, day)

    @staticmethod
    def difference_in_months(date1: datetime, date2: datetime) -> int:
        """Calculate difference in months between two dates"""
        return (date1.year - date2.year) * 12 + date1.month - date2.month

    @staticmethod
    def format_date(date: datetime) -> str:
        """Format date for display"""
        return date.strftime("%b %d, %Y")

    @classmethod
    def calculate_payment_strategy(cls, data: Dict[str, Any]) -> Results:
        """Calculate the optimal payment strategy"""
        total_amount = float(data.get("totalAmount", 0))
        installments = int(data.get("installments", 0))
        on_time_fee = float(data.get("onTimeFee", 0))
        return_rate = float(data.get("returnRate", 0))
        first_date = datetime.strptime(data.get("firstPaymentDate"), "%Y-%m-%d")
        lump_sum_date = datetime.strptime(data.get("lumpSumDate"), "%Y-%m-%d")
        postpone_months = int(data.get("postponeMonths", 0))

        installment_amount = total_amount / installments
        max_postpone_months = cls.difference_in_months(lump_sum_date, first_date)
        postpone_months = min(postpone_months, max_postpone_months)

        # Generate payment schedule with postponement
        payment_schedule = []
        for i in range(installments-postpone_months):
            payment_date = cls.add_months(first_date, i + postpone_months)
            amount = installment_amount
            fee = 0
            accumulated_payments = 1

            # Handle postponement logic
            if i == 0 and postpone_months > 0:
                # 1st payment after postponement has all postponed payments
                accumulated_payments = postpone_months + 1
                amount = installment_amount * accumulated_payments
                # One-time fee applied to first actual payment
                fee = on_time_fee
            elif i == 0:
                # Regular first payment
                fee = on_time_fee

            remaining_balance = total_amount - (installment_amount * (i + 1))

            schedule_item = PaymentScheduleItem(
                payment_number=i + 1,
                date=cls.format_date(payment_date),
                amount=amount,
                fee=fee,
                remaining_balance=max(0, remaining_balance),
            )

            # Add accumulated payments info if applicable
            if accumulated_payments > 1:
                schedule_item.accumulated_payments = accumulated_payments

            payment_schedule.append(schedule_item)

        # Calculate installment strategy
        total_fees = on_time_fee

        # Calculate returns for remaining balance with monthly compounding
        # For postponed payments, investment returns accrue during
        # postponement and after each payment
        investment_returns = 0
        monthly_return_rate = return_rate / 100 / 12  # Monthly return rate
        investment_breakdown = []

        # Investment returns during postponement period
        for i in range(postpone_months):
            monthly_return = total_amount * monthly_return_rate
            investment_returns += monthly_return
            investment_breakdown.append(
                InvestmentBreakdownItem(
                    month=i + 1,
                    balance=total_amount,
                    monthly_return=monthly_return,
                    cumulative_return=investment_returns,
                )
            )

        # Investment returns after each payment (excluding the final payment)
        for i in range(postpone_months, installments - 1):
            remaining_balance = total_amount - (installment_amount * (i + 1))
            if remaining_balance > 0:
                monthly_return = remaining_balance * monthly_return_rate
                investment_returns += monthly_return
                investment_breakdown.append(
                    InvestmentBreakdownItem(
                        month=i + 1,
                        balance=remaining_balance,
                        monthly_return=monthly_return,
                        cumulative_return=investment_returns,
                    )
                )

        installment_net_cost = total_amount + total_fees - investment_returns

        # Calculate lump sum strategy
        months_to_lump_sum = cls.difference_in_months(lump_sum_date, first_date)
        opportunity_cost = (
            total_amount * (return_rate / 100) * (months_to_lump_sum / 12)
        )
        lump_sum_net_cost = total_amount + opportunity_cost

        # Determine recommendation
        savings = abs(installment_net_cost - lump_sum_net_cost)
        if lump_sum_net_cost < installment_net_cost:
            strategy = "Lump Sum"
        else:
            strategy = "Installment"

        return Results(
            recommendation={"strategy": strategy, "savings": savings},
            installment={
                "totalAmount": total_amount,
                "totalFees": total_fees,
                "investmentReturns": investment_returns,
                "netCost": installment_net_cost,
            },
            lump_sum={
                "totalAmount": total_amount,
                "fees": 0,
                "opportunityCost": opportunity_cost,
                "netCost": lump_sum_net_cost,
            },
            payment_schedule=payment_schedule,
            investment_breakdown=investment_breakdown,
        )